### PR TITLE
perf(bridge-history): optimize get claimable l2 sent msg query

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

This PR boosts the speed of `GetClaimableL2SentMsgByAddressWithOffset` and `GetClaimableL2SentMsgByAddressTotalNum`. Key changes:

Combine Functions: Merges both functions into one that returns total unclaimed transactions and a paginated list.

Simplify Queries: Adapts query logic assuming fewer transactions per address. The new approach:

i. Fetches all transactions for the address.
ii. Identifies claimed transactions.
iii. Derives unclaimed transactions.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] perf: A code change that improves performance

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [ ] No, this PR doesn't involve a new deployment, git tag, docker image tag
- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: bridge-history-api/orm/l2_sent_msg.go [hunk 1]
```diff
@@ -73,25 +73,52 @@ func (l *L2SentMsg) GetLatestSentMsgHeightOnL2(ctx context.Context) (uint64, err
...
+	db = db.Where("original_sender = ? OR sender = ?", address, address)
+	db = db.Where("msg_proof != ''")
+	db = db.Where("deleted_at IS NULL")
+	if err := db.Find(&totalMsgs).Error; err != nil {
+		return 0, nil, err
+	}
+
+	msgHashes := make([]string, len(totalMsgs))
+	for i, msg := range totalMsgs {
+		msgHashes[i] = msg.MsgHash
```
Comment (Thegaram) [inline bridge-history-api/orm/l2_sent_msg.go]: Why use `MsgHash` and not `Id`?

File: bridge-history-api/orm/l2_sent_msg.go [hunk 2]
```diff
@@ -73,25 +73,53 @@ func (l *L2SentMsg) GetLatestSentMsgHeightOnL2(ctx context.Context) (uint64, err
...
+	}
+
+	var claimedMsgHahes []string
+	db = l.db.WithContext(ctx)
+	db = db.Table("relayed_msg")
+	db = db.Where("msg_hash IN ?", msgHashes)
```
Comment (Thegaram) [inline bridge-history-api/orm/l2_sent_msg.go]: They say `IN` is quite slow with >100 values, and there are some alternatives like this https://postgres.cz/wiki/PostgreSQL_SQL_Tricks_I#Predicate_IN_optimalization but I'm not deeply knowledgeable about this.

File: bridge-history-api/orm/l2_sent_msg.go [hunk 3]
```diff
@@ -73,25 +73,53 @@ func (l *L2SentMsg) GetLatestSentMsgHeightOnL2(ctx context.Context) (uint64, err
...
+	if err := db.Find(&totalMsgs).Error; err != nil {
+		return 0, nil, err
+	}
+
+	msgHashes := make([]string, len(totalMsgs))
+	for i, msg := range totalMsgs {
+		msgHashes[i] = msg.MsgHash
+	}
+
+	var claimedMsgHahes []string
```
Comment (Thegaram) [inline bridge-history-api/orm/l2_sent_msg.go]:
```suggestion
	var claimedMsgHashes []string
```

File: bridge-history-api/orm/l2_sent_msg.go (hunk 2 above)
Comment (Thegaram) [reply bridge-history-api/orm/l2_sent_msg.go]: We should add a metric about these result set sizes in the future.

**Review by ChuhanJin — DISMISSED** [review]

File: bridge-history-api/orm/l2_sent_msg.go (hunk 1 above)
Comment (Thegaram) [reply bridge-history-api/orm/l2_sent_msg.go]: I see now, it's because you're correlating the msg hashes from two tables.

**Review by Thegaram — DISMISSED** [review]

File: bridge-history-api/orm/l2_sent_msg.go (hunk 1 above)
Comment (colinlyguo) [reply bridge-history-api/orm/l2_sent_msg.go]: Yes.

File: bridge-history-api/orm/l2_sent_msg.go (hunk 3 above)
Comment (colinlyguo) [reply bridge-history-api/orm/l2_sent_msg.go]: changed in [75bdd60](https://github.com/scroll-tech/scroll/pull/959/commits/75bdd606dca809d307de6dfe8cd2d13856b47273)

File: bridge-history-api/orm/l2_sent_msg.go (hunk 2 above)
Comment (colinlyguo) [reply bridge-history-api/orm/l2_sent_msg.go]: added a comment about metric in [75bdd60](https://github.com/scroll-tech/scroll/pull/959/commits/75bdd606dca809d307de6dfe8cd2d13856b47273)
and changed query style to `VALUES(10),(20)`.

Comment (Thegaram) [reply bridge-history-api/orm/l2_sent_msg.go]: Thanks. This syntax looks a bit unusual, can you test that it works before we merge this PR?

Comment (colinlyguo) [reply bridge-history-api/orm/l2_sent_msg.go]: ~~It does not work (`db = db.Where(fmt.Sprintf("msg_hash IN (VALUES %s)", valuesStr))`), so I change it back.~~

Comment (colinlyguo) [reply bridge-history-api/orm/l2_sent_msg.go]: fixed in [1ea57c4](https://github.com/scroll-tech/scroll/pull/959/commits/1ea57c464aeca67e7b2273da207020c5b19c4323)

File: bridge-history-api/orm/l2_sent_msg.go [hunk 4]
```diff
@@ -72,26 +73,57 @@ func (l *L2SentMsg) GetLatestSentMsgHeightOnL2(ctx context.Context) (uint64, err
...
+func (l *L2SentMsg) GetClaimableL2SentMsgByAddressWithOffset(ctx context.Context, address string, offset int, limit int) (uint64, []*L2SentMsg, error) {
+	var totalMsgs []*L2SentMsg
+	db := l.db.WithContext(ctx)
+	db = db.Table("l2_sent_msg")
+	db = db.Where("original_sender = ? OR sender = ?", address, address)
+	db = db.Where("msg_proof != ''")
+	db = db.Where("deleted_at IS NULL")
+	db = db.Order("id DESC")
+	tx := db.Find(&totalMsgs)
+	if tx.Error != nil || tx.RowsAffected == 0 {
```
Comment (Thegaram) [inline bridge-history-api/orm/l2_sent_msg.go:87]: Is `gorm.ErrRecordNotFound` handled elsewhere?

File: bridge-history-api/orm/l2_sent_msg.go [hunk 5]
```diff
@@ -72,26 +73,57 @@ func (l *L2SentMsg) GetLatestSentMsgHeightOnL2(ctx context.Context) (uint64, err
...
+		return 0, nil, tx.Error
+	}
+
+	var valuesStr string
+	for _, msg := range totalMsgs {
+		valuesStr += fmt.Sprintf("('%s'),", msg.MsgHash)
```
Comment (Thegaram) [inline bridge-history-api/orm/l2_sent_msg.go:106]: It's worth adding a comment about why we do this.

**Review by Thegaram — DISMISSED** [review]

File: bridge-history-api/orm/l2_sent_msg.go (hunk 4 above)
Comment (colinlyguo) [reply bridge-history-api/orm/l2_sent_msg.go:87]: this syntax would not return `gorm.ErrRecordNotFound`.

File: bridge-history-api/orm/l2_sent_msg.go (hunk 5 above)
Comment (colinlyguo) [reply bridge-history-api/orm/l2_sent_msg.go:106]: a comment is added in [1f8d534](https://github.com/scroll-tech/scroll/pull/959/commits/1f8d534660aa072d5f80637627b0fa762c4af4f9)

**Review by Thegaram — APPROVED** [review]

**Review by georgehao — APPROVED** [review]

**Pull request merged**